SIG Proposal for Sustainability Metrics (#2018)
open-telemetry/community: projects/env-semconv.md

### Description

This working group looks to identify and standardize observability for
sustainability purposes with regards to infrastructure and applications. This
would look at current tooling around sustainability metrics and standardize the
meaning and approach for these metrics.

Current efforts of implementations:

- [Kepler](https://github.com/sustainable-computing-io/kepler)
- [Scaphandre](https://github.com/hubblo-org/scaphandre)

> We want to have a unified metrics convention in order to better facilitate measuring sustainability metrics across different ecosystems

Reviewer: ```suggestion
We want to have a unified metrics convention in order to better facilitate measuring sustainability metrics across different ecosystems.
```

### Deliverables

- Semantic conventions for sustainability metrics
- Reference Prototypes metrics

> - Recommended practices for reporting on sustainability metrics 

Reviewer 2: I see "recommended practices" is mentioned here, but not in the timeline. Can this be scoped a little tighter? Is this related to "how" to measure or "what/where" to measure for instance? What form will these recommendations be given as?

Author: I think this is part of the discussion to be had. I would go to "the recommended way to measure a metric", but think thats up for discussion

Reviewer: Thanks for putting this on the time line.

Given that there are many misconceptions and wrong assumptions around "sustainability", those recommended practices could also offer some guidelines for correctly interpreting those metrics.

Reviewer 2: That makes sense. I just wanted to discuss the potential scope creep in terms of the guidance that OpenTelemetry can give. For instance, what sustainability metrics each organisation should pay more attention to, depending on their workloads and data centre placement, or how to interpret them, is something that other groups are probably better place at answering. However, as that will be discussed as part of this group I'm happy with the answer "within the scope of OpenTelemetry".

### Staffing / Help Wanted

**Project Leads**
- @spazzy757

**Sponsoring Members**
- @jpkrohling 
- @pyohannes 

**Engineers**
- @gabibeyer
- @bertysentry
- @nikimanoledaki
- @mkorbi
- @rossf7
- @marceloamaral

### Meeting Times

- Meeting Times: Mondays 4:00PM - 5:00PM CET

> - CNCF Slack [#otel-sustainability-semconv-wg](https://cloud-native.slack.com/archives/C06RTM63YD6) for collaboration

Reviewer 3: (#2044)

```suggestion
- CNCF Slack [#otel-sustainability-semconv-sig](https://cloud-native.slack.com/archives/C06RTM63YD6) for collaboration
```

### Timeline

> 1. Initial Draft end of Q3 2024
> 1. Review and Refinement: TBD
> 1. Prototypes: TBD

Reviewer: ```suggestion
Q3 2024

* Kickoff
* Experimental semantic conventions

Q4 2024

* Prototype development

1H 2025

* Review and refinement
* Declare sustainability semantic conventions as stable
```

@Spazzy757 let me know if that makes sense to you. It's better to have a very rough outline and order instead of TBDs. There's nothing cut in stone though, we can adapt as we go.

### Description

This working group looks to identify and standardize observability for
sustainability purposes with regards to infrastructure and applications. This
would look at current tooling around sustainability metrics and standardize the
meaning and approach for these metrics.

Current efforts of implementations:

- [Kepler](https://github.com/sustainable-computing-io/kepler)
- [Scaphandre](https://github.com/hubblo-org/scaphandre)
- TBD

We want to have a unified metrics convention in order to better facilitate measuring sustainability metrics across different ecosystems

### Deliverables

- Semantic conventions for sustainability metrics
- Reference Prototypes metrics
- Recommended practices for reporting on sustainability metrics 

### Staffing / Help Wanted

**Project Leads**
- @spazzy757

**Sponsoring Members**

> - @jpkrohling 

Reviewer 3: I confirmed that @pyohannes can be added here as a sponsoring member, thanks @pyohannes!

### Description

This working group looks to identify and standardize observability for
sustainability purposes with regards to infrastructure and applications. This
would look at current tooling around sustainability metrics and standardize the
meaning and approach for these metrics.

Current efforts of implementations:

- [Kepler](https://github.com/sustainable-computing-io/kepler)
- [Scaphandre](https://github.com/hubblo-org/scaphandre)

> - TBD

Reviewer: Do we expect more entries here? Otherwise this could be removed.

We want to have a unified metrics convention in order to better facilitate measuring sustainability metrics across different ecosystems

### Deliverables

- Semantic conventions for sustainability metrics
- Reference Prototypes metrics
- Recommended practices for reporting on sustainability metrics 

### Staffing / Help Wanted

**Project Leads**
- @spazzy757

**Sponsoring Members**
- @jpkrohling 
- @pyohannes 

**Engineers**
- @gabibeyer
- @bertysentry
- @nikimanoledaki
- @mkorbi
- @rossf7

> - @marceloamaral

Reviewer: This should be added to the `.cspell.yaml`.

**Maintainers**
- To be determined

> 

Reviewer: This can be removed.

### Meeting Times

- Meeting Times: TBD

> - [cncf slack channel](https://cloud-native.slack.com/archives/C06RTM63YD6) for collaboration

Reviewer: ```suggestion
- CNCF Slack [#otel-sustainability-semconv-wg](https://cloud-native.slack.com/archives/C06RTM63YD6) for collaboration
```

### Description

This working group looks to identify and standardize observability for
sustainability purposes with regards to infrastructure and applications. This
would look at current tooling around sustainability metrics and standardize the
meaning and approach for these metrics.

Current efforts of implementations:

- [Kepler](https://github.com/sustainable-computing-io/kepler)
- [Scaphandre](https://github.com/hubblo-org/scaphandre)

We want to have a unified metrics convention in order to better facilitate measuring sustainability metrics across different ecosystems.

### Deliverables

- Semantic conventions for sustainability metrics
- Reference Prototypes metrics
- Recommended practices for reporting on sustainability metrics 

### Staffing / Help Wanted

**Project Leads**
- @spazzy757

**Sponsoring Members**
- @jpkrohling 
- @pyohannes 

**Engineers**
- @gabibeyer
- @bertysentry
- @nikimanoledaki
- @mkorbi
- @rossf7
- @marceloamaral

### Meeting Times

- Meeting Times: Mondays 4:00PM - 5:00PM CET
- CNCF Slack [#otel-sustainability-semconv-sig](https://cloud-native.slack.com/archives/C06RTM63YD6) for collaboration

### Timeline

> Q3 2024
> * Kickoff
> * Experimental semantic conventions
> Q4 2024
> * Prototype development
> 1H 2025
> * Review and refinement
> * Declare sustainability semantic conventions as stable
> * Create Recommendations

Reviewer: ```suggestion
Q3 2024

* Kickoff
* Experimental semantic conventions

Q4 2024

* Prototype development

1H 2025

* Review and refinement
* Declare sustainability semantic conventions as stable
* Create Recommendations
```

Otherwise it's not formatted correctly.

### Description

This working group looks to identify and standardize observability for
sustainability purposes with regards to infrastructre and applications. This

> would look at current tooling around sustainability metrics and standardize the
> meaning and approach for these metrics

Reviewer: ```suggestion
would look at current tooling around sustainability metrics and standardize the
meaning and approach for these metrics.
```

> 

Reviewer: ```suggestion
```

> current efforts of implementations:

Reviewer: ```suggestion
Current efforts of implementations:
```

- [Kepler](https://github.com/sustainable-computing-io/kepler)

> - [Scaphandre](https://github.com/hubblo-org/scaphandre)

Reviewer: Would people from these projects be willing to be involved here? Do you know if they would be willing to standardize on OTel semantic conventions?

Author: We can talk with them, and find out, they do have a relationship with TAG ENV

Author: It the Kepler project is interested, I'll push for someone to maybe join

Reviewer 4: @pyohannes and @Spazzy757 this could be indeed interesting from the Kepler standpoint

However, some time ago, there was a [presentation during the Kepler community](https://github.com/danielm0hr/edge-metrics-measurements/blob/main/talks/DanielMohr_PromAgentVsOtelCol.pdf) that included an overhead comparison between the OpenTelemetry (OTel) and Prometheus SDKs, showing that the OTel SDK had significantly higher overhead.

As you know, the key difference between these SDKs is how they export metrics. With OpenTelemetry, instrumentation libraries within the application actively push data to backend systems, whereas the Prometheus SDK relies on applications to expose HTTP endpoints for scraping by a monitoring server. These approaches represent orthogonal push and pull approaches.

However, while the Prometheus SDK does offer the option to [enable the OpenMetrics format](https://github.com/prometheus/client_golang/blob/main/examples/exemplars/main.go#L64), it still follows a pull-based approach. 

Given that performance is very important to us, my questions are: do we really need to migrate from Prometheus SDK to OTel SDK? Has the overhead of the OpenTelemetry SDK been improved?

Reviewer 5: Hi @marceloamaral, I just wanted to mention that, in those experiments shared in the community, I don't think the OTel SDKs were used. (Please, @danielm0hr, correct me if I'm wrong). Metrics are produced with avalanche in prometheus format and, then, scraped with different mechanisms. (Prometheus, or OpenTelemetry collector in different setups).

My main takeaway from this work is that, when collecting prometheus metrics, (by the time the experiments were made) it was more efficient to just use the prometheus tooling. I don't have proof if this would be the sole contribution for the overhead in usage but my suspicions are pointing to the fact that the OpenTelemetry collector needed to transform from prometheus to OpenTelemetry formats back and forth in this experiment (which is still very valid on its own, but I don't think it will apply for the SDK performance discussion).

For the (very relevant, and important) points that you are raising for assessing the viability of changing the SDK, I think the comparison should be between: 1) metrics generated with the OTel SDK (with the OTel pipeline) vs  2) the current prometheus approach. 

BTW, changing to the OpenTelemetry SDK should not be an impediment to continue using prometheus metrics ( see [prometheus exporter](https://opentelemetry.io/docs/languages/go/exporters/#prometheus-experimental) ).

I hope this helps a bit! We can also follow up in the Kepler community if needed!

Reviewer 6: The proposal of this issue is also not related to the SDKs, but to agree on a common convention for sustainability-related metrics.

Reviewer 4: @jgomezselles, thanks for the clarification! It will indeed be interesting to have another experiment comparing the overhead of the OTel versus Prometheus SDK. Any volunteer? :)

Reviewer 4: @dmathieu, right, that makes sense! 
I'd be happy to participate in the discussion.

Author: @marceloamaral I have added you to the list of engineers

### Description

This working group looks to identify and standardize observability for
sustainability purposes with regards to infrastructure and applications. This
would look at current tooling around sustainability metrics and standardize the
meaning and approach for these metrics.

Current efforts of implementations:

- [Kepler](https://github.com/sustainable-computing-io/kepler)
- [Scaphandre](https://github.com/hubblo-org/scaphandre)

> We want to have a unified metrics convention in order to better facilitate measuring sustainability metrics across different ecosystems.

Reviewer 7: Just a note/quibble I guess, but I think you should be careful with 'metrics' vs. 'Metrics' in the OpenTelemetry sense. For example, there is no reason that various sustainability metrics can't be attached to other data types in OpenTelemetry (e.g., emitting some sort of co2 usage as a span attribute on a given FaaS function execution).

### Deliverables

- Semantic conventions for sustainability metrics
- Reference Prototypes metrics
- Recommended practices for reporting on sustainability metrics within the scope of OpenTelemetry 

### Staffing / Help Wanted

**Project Leads**
- @spazzy757

**Sponsoring Members**

> - @jpkrohling 

Reviewer 2: I can take this @jpkrohling (there's a requirement for sponsors to be from separate companies).
